fcrepo4-exts/migration-utils
# Allow selection of PIDs/Objects to migrate with to new selectors

1. ResumePidList
   - Allows resuming the migration from the point of the last successfully migration F3 object
2. UserProvidedPidList
   - Allows for user to provide a list of PIDs to migrate

Resolves:
https://jira.duraspace.org/browse/FCREPO-3111 , and
https://jira.duraspace.org/browse/FCREPO-3112

# What's new?
Two new "PidListManagers" have been created. They can be configured in the spring.xml as follows:
```
    <bean id="migrator" class="org.fcrepo.migration.Migrator">
        <property name="source" ref="${migration.layout:exported}" />
        <property name="handler" ref="objectAbstraction" />
        <property name="limit" value="${migration.limit:2}" />
        
        <property name="pidListManagers">
            <list>
                <bean class="org.fcrepo.migration.pidlist.UserProvidedPidListManager">
                    <constructor-arg name="pidListFile" value="${migration.pid.list.file:#{null}}"/>
                </bean>
                <bean class="org.fcrepo.migration.pidlist.ResumePidListManager">
                    <constructor-arg name="pidDir" value="${migration.pid.resume.dir:target/test/pid}"/>
                    <constructor-arg name="acceptAll" value="${migration.pid.resume.all:true}"/>
                </bean>
            </list>
        </property>
    </bean>
```
The idea is that these two "PidListManagers" will be executed in order, each being asked if they "accept" the provided PID. If either of the managers do not "accept" the PID, the PID will not be processed.
The `UserProvidedPidListManager` takes a file of line-separated PIDs to process. If no file is provided, all PIDs are accepted.
The `ResumePidListManager` keeps track of the PID and index of the last processed object. If `acceptAll` is true, this state of "last processed object" is maintained, but all PIDs are "accepted". However, if `acceptAll` is false, any PIDs that come before the PID/index found in the state file, will not be processed. Note, this assumes a deterministic, repeatable order of source PIDs.

# How should this be tested?
Besides the unit and integration tests provided in this pull-request, testing should involve:
- Configuring the spring.xml with one, both or none of the "PidListManagers"
- Run, kill and re-run the tool, to test the "resume" functionality
- Create a pid-list to test the pid-list functionality

# Interested parties
@fcrepo4/committers
